File: AccuPay/Domain/TimeEntry/AgencyFeeCalculator.vb
Option Strict On

Imports AccuPay
Imports AccuPay.Entity
Imports AccuPay.Utils

Public Class AgencyFeeCalculator

    Private Const HoursPerDay As Decimal = 8

    Private ReadOnly _employee As Employee
    Private ReadOnly _agency As Agency
    Private ReadOnly _agencyFees As IList(Of AgencyFee)

    Public Sub New(employee As Employee, agency As Agency, agencyFees As IList(Of AgencyFee))
        _employee = employee
        _agency = agency
        _agencyFees = agencyFees
    End Sub

    Public Function Compute(timeEntries As IList(Of TimeEntry)) As IList(Of AgencyFee)
        Dim agencyFees = New List(Of AgencyFee)

        For Each timeEntry In timeEntries
            Dim agencyFee = _agencyFees.SingleOrDefault(Function(a) a.Date = timeEntry.Date)

            If agencyFee Is Nothing Then
                agencyFee = New AgencyFee() With {
                    .OrganizationID = _employee.OrganizationID,
                    .EmployeeID = _employee.RowID,
                    .AgencyID = _agency.RowID,
                    .Date = timeEntry.Date}
            End If

            Dim basicHours = AccuMath.CommercialRound(timeEntry.BasicHours)
            Dim hourlyFee = _agency.Fee / HoursPerDay
            agencyFee.Amount = basicHours * hourlyFee

            agencyFees.Add(agencyFee)
        Next

        Return agencyFees
    End Function

End Class

File: Accupay.Data/Services/TimeEntryGeneration/AgencyFeeCalculator.cs
using AccuPay.Data.Entities;
using AccuPay.Utilities;
using System.Collections.Generic;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class AgencyFeeCalculator
    {
        private const decimal HoursPerDay = 8;

        private readonly Employee _employee;
        private readonly Agency _agency;
        private readonly IList<AgencyFee> _agencyFees;

        public AgencyFeeCalculator(Employee employee, Agency agency, IList<AgencyFee> agencyFees)
        {
            _employee = employee;
            _agency = agency;
            _agencyFees = agencyFees;
        }

        public IList<AgencyFee> Compute(IList<TimeEntry> timeEntries)
        {
            var agencyFees = new List<AgencyFee>();

            foreach (var timeEntry in timeEntries)
            {
                var agencyFee = _agencyFees.SingleOrDefault(a => a.Date == timeEntry.Date);

                if (agencyFee == null)
                {
                    agencyFee = new AgencyFee()
                    {
                        OrganizationID = _employee.OrganizationID,
                        EmployeeID = _employee.RowID,
                        AgencyID = _agency.RowID,
                        Date = timeEntry.Date
                    };
                }

                var basicHours = AccuMath.CommercialRound(timeEntry.BasicHours);
                var hourlyFee = _agency.Fee / HoursPerDay;
                agencyFee.Amount = basicHours * hourlyFee;

                agencyFees.Add(agencyFee);
            }

            return agencyFees;
        }
    }
}